File: sangi_practice/sangi_practice_12/model_answers/cpp/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n, w;
    cin >> n >> w;
    vector<int> dp(w + 1, 0);
    for (int i = 0; i < n; i++) {
        int cost, value;
        cin >> cost >> value;
        for (int j = w; j >= cost; j--) {
            dp[j] = max(dp[j], dp[j - cost] + value);
        }
    }
    cout << dp[w] << "\n";
    return 0;
}

File: sangi_practice/sangi_practice_12/model_answers/java/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), w = sc.nextInt();
        int[] dp = new int[w + 1];
        for (int i = 0; i < n; i++) {
            int cost = sc.nextInt(), value = sc.nextInt();
            for (int j = w; j >= cost; j--) {
                dp[j] = Math.max(dp[j], dp[j - cost] + value);
            }
        }
        System.out.println(dp[w]);
    }
}
